src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStreamAsyncMemoryOverloads.cs

```diff
@@ -0,0 +1,133 @@
+                                                                                        isDataflowRule: false);
+
+        internal static DiagnosticDescriptor PreferStreamWriteAsyncMemoryOverloadsRule = DiagnosticDescriptorHelper.Create(
+                                                                                        RuleId,
+                                                                                        s_localizableTitleWrite,
+                                                                                        s_localizableMessageWrite,
+                                                                                        DiagnosticCategory.Performance,
+                                                                                        RuleLevel.IdeSuggestion,
+                                                                                        s_localizableDescriptionWrite,
+                                                                                        isPortedFxCopRule: false,
+                                                                                        isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(PreferStreamReadAsyncMemoryOverloadsRule, PreferStreamWriteAsyncMemoryOverloadsRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(AnalyzeCompilationStart);
+        }
+
+        private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemMemory1) == null ||
+            !context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIOStream, out INamedTypeSymbol? stream))
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context =>
+            {
+                IInvocationOperation invocation = (IInvocationOperation)context.Operation;
+                IMethodSymbol method = invocation.TargetMethod;
+
+                DiagnosticDescriptor rule;
+                if (string.Equals(method.Name, "ReadAsync", StringComparison.Ordinal))
+                {
+                    rule = PreferStreamReadAsyncMemoryOverloadsRule;
+                }
+                else if (string.Equals(method.Name, "WriteAsync", StringComparison.Ordinal))
+                {
+                    rule = PreferStreamWriteAsyncMemoryOverloadsRule;
+                }
+                else
+                {
+                    return;
+                }
+
+                if (IsStreamMethod(method, stream) && HasUndesiredArguments(method))
```

Comment:
I think we should only warn when the call's result is directly awaited.  Changing which overload is used will change the return type, so as written this will warn (and a code fixer will break) code like:
```C#
Task t = stream.ReadAsync(buffer, 0, buffer.Length);
```
or
```C#
public Task DoWorkAsync() => _stream.WriteAsync(_buffer, 0, _buffer.Length);
```

Reply:
Good catch, thanks. I'll work on this case.

Reply:
I'm now handling this case in my latest commit. I also added unit tests.